API-ify Datagen

ChiefArug opened the pull request:
This refactors datagen into an API that addons can make use of without the overhead of also running all of Creates stuff.

This splits each current datagen class into a base class in the api package, which contains things like the recipe type, the generation logic and most of the helper methods, and a final class in Create's data package which contains all the actual generation and any helper methods that make use of the Mod enum (my original plan was to weaken that to package-private as it is inextensible and therefore bad for addons to be using, however despite being in the data package it is used in many places outside that for hardcoded mod compat).

Fixes #7366
Conflicts with #7084, one of these will need editing after the other gets merged.

ChiefArug committed:
Refactor datagen so that addons have a clear API to follow.

ChiefArug committed:
Run datagen, only changed files are a couple caches where the display name changed.

ChiefArug committed:
Clean up

ChiefArug committed:
Merge remote-tracking branch 'origin/mc1.20.1/dev' into 1.20.1

# Conflicts:
#	src/main/java/com/simibubi/create/foundation/data/recipe/CreateCuttingRecipeGen.java

ChiefArug commented:
This is a massive breaking change, however it shouldn't actually break addons at runtime, only in dev.

StaticFX commented:
I really like this, and would also offer to write docs for this

ChiefArug commented:
Oh, yeah.. docs. Those might be a good thing to have. 

> would also offer to write docs for this


If you want you can open a pr to this branch on my fork so I can merge it into this pr, or failing that I could write them sometime in the next week (I'd rather not, but I've been stung by the lack of docs before so would do ie if need be), or if Create is okay with this being merged without docs you could do a seperate PR to Create after this is merged.

StaticFX commented:
I dont really mind tbh, would first like ofc get some feedback if this is even going to get merged :D. Ive also worked on my own API Ficication for the Processing API. My goal is, to make a doc entry, and also add javadocs to make it clear what to do. And then when this is done, expand it further in the future to other parts of create mod.

ChiefArug committed:
Merge branch '1.20.1' into apify-datagen

ChiefArug committed:
Fix errors introduced in merging.

ChiefArug committed:
Add pointer javadocs to all of Create's own datagen classes pointing to the API classes.

ChiefArug committed:
Access GeneratedRecipe directly rather than through a subclass
 Probably a weird artifact from me moving files about and IntelliJ's automatic refactoring.

ChiefArug committed:
Add javadoc to Mods.java explaining its non-API status

ChiefArug committed:
Begin Javadocs for API section.
Also includes small modification to CreateRecipeProvider.java to help people work up the tree of classes used for datagen.

ChiefArug commented:
I have begun with javadocs, feedback is appreciated as I have never actually used datagen (Create's or otherwise), this is just from my understanding from reading the code whilst making this PR.

ChiefArug committed:
Some cleanup
Remove SuppressWarnings annotation on datagen api classes and make all datagen api classes abstract.

ChiefArug committed:
Fully document MechanicalCraftingRecipeBuilder.java and give its parameters proper names.

ChiefArug committed:
Add class javadocs to all RecipeGen classes
Also update CreateRecipeProvider's javadoc as it does not generate all of Create's recipes for some reason.

IThundxr commented:
Thanks!
